File: src/genesis_ts/src/dna-lang/lexer.ts
/**
 * DNA-Lang Lexer
 *
 * Tokenizes DNA-Lang source code into tokens for parsing.
 */

export enum TokenType {
  // Keywords
  ORGANISM = 'ORGANISM',
  GENE = 'GENE',
  TRAIT = 'TRAIT',
  EXPRESS = 'EXPRESS',
  MUTATE = 'MUTATE',
  EVOLVE = 'EVOLVE',
  MEMBRANE = 'MEMBRANE',
  NUCLEUS = 'NUCLEUS',
  MITOCHONDRIA = 'MITOCHONDRIA',
  
  // Operators
  BIND = 'BIND',         // }{
  TRANSCRIBE = 'TRANSCRIBE', // ->
  TRANSLATE = 'TRANSLATE',   // =>
  REPLICATE = 'REPLICATE',   // ><
  SPLICE = 'SPLICE',     // |
  CODON = 'CODON',       // :
  HELIX = 'HELIX',       // ~
  
  // Literals
  IDENTIFIER = 'IDENTIFIER',
  NUMBER = 'NUMBER',
  STRING = 'STRING',
  
  // Delimiters
  LPAREN = 'LPAREN',     // (
  RPAREN = 'RPAREN',     // )
  LBRACE = 'LBRACE',     // {
  RBRACE = 'RBRACE',     // }
  LBRACKET = 'LBRACKET', // [
  RBRACKET = 'RBRACKET', // ]
  COMMA = 'COMMA',       // ,
  SEMICOLON = 'SEMICOLON', // ;
  
  // Special
  COMMENT = 'COMMENT',
  NEWLINE = 'NEWLINE',
  EOF = 'EOF',
  ERROR = 'ERROR',
}

export interface Token {
  type: TokenType;
  value: string;
  line: number;
  column: number;
}

const KEYWORDS: Record<string, TokenType> = {
  'organism': TokenType.ORGANISM,
  'gene': TokenType.GENE,
  'trait': TokenType.TRAIT,
  'express': TokenType.EXPRESS,
  'mutate': TokenType.MUTATE,
  'evolve': TokenType.EVOLVE,
  'membrane': TokenType.MEMBRANE,
  'nucleus': TokenType.NUCLEUS,
  'mitochondria': TokenType.MITOCHONDRIA,
};

export class Lexer {
  private source: string;
  private pos: number = 0;
  private line: number = 1;
  private column: number = 1;
  private tokens: Token[] = [];

  constructor(source: string) {
    this.source = source;
  }

  /**
   * Tokenize the entire source.
   */
  tokenize(): Token[] {
    while (!this.isAtEnd()) {
      this.scanToken();
    }
    this.tokens.push({
      type: TokenType.EOF,
      value: '',
      line: this.line,
      column: this.column,
    });
    return this.tokens;
  }

  private isAtEnd(): boolean {
    return this.pos >= this.source.length;
  }

  private peek(): string {
    if (this.isAtEnd()) return '\0';
    return this.source[this.pos];
  }

  private peekNext(): string {
    if (this.pos + 1 >= this.source.length) return '\0';
    return this.source[this.pos + 1];
  }

  private advance(): string {
    const ch = this.source[this.pos];
    this.pos++;
    if (ch === '\n') {
      this.line++;
      this.column = 1;
    } else {
      this.column++;
    }
    return ch;
  }

  private addToken(type: TokenType, value: string): void {
    this.tokens.push({
      type,
      value,
      line: this.line,
      column: this.column - value.length,
    });
  }

  private scanToken(): void {
    const ch = this.advance();

    switch (ch) {
      // Single character tokens
      case '(': this.addToken(TokenType.LPAREN, ch); break;
      case ')': this.addToken(TokenType.RPAREN, ch); break;
      case '[': this.addToken(TokenType.LBRACKET, ch); break;
      case ']': this.addToken(TokenType.RBRACKET, ch); break;
      case ',': this.addToken(TokenType.COMMA, ch); break;
      case ';': this.addToken(TokenType.SEMICOLON, ch); break;
      case ':': this.addToken(TokenType.CODON, ch); break;
      case '~': this.addToken(TokenType.HELIX, ch); break;
      case '|': this.addToken(TokenType.SPLICE, ch); break;

      // Two-character tokens
      case '}':
        if (this.peek() === '{') {
          this.advance();
          this.addToken(TokenType.BIND, '}{');
        } else {
          this.addToken(TokenType.RBRACE, ch);
        }
        break;

      case '{':
        this.addToken(TokenType.LBRACE, ch);
        break;

      case '-':
        if (this.peek() === '>') {
          this.advance();
          this.addToken(TokenType.TRANSCRIBE, '->');
        } else {
          // Might be part of a number
          if (this.isDigit(this.peek())) {
            this.scanNumber('-');
          } else {
            this.addToken(TokenType.ERROR, ch);
          }
        }
        break;

      case '=':
        if (this.peek() === '>') {
          this.advance();
          this.addToken(TokenType.TRANSLATE, '=>');
        } else {
          this.addToken(TokenType.ERROR, ch);
        }
        break;

      case '>':
        if (this.peek() === '<') {
          this.advance();
          this.addToken(TokenType.REPLICATE, '><');
        } else {
          this.addToken(TokenType.ERROR, ch);
        }
        break;

      // Comments
      case '#':
        this.scanComment();
        break;

      // Whitespace
      case ' ':
      case '\r':
      case '\t':
        // Ignore whitespace
        break;

      case '\n':
        this.addToken(TokenType.NEWLINE, ch);
        break;

      // String literals
      case '"':
        this.scanString();
        break;

      default:
        if (this.isDigit(ch)) {
          this.scanNumber(ch);
        } else if (this.isAlpha(ch)) {
          this.scanIdentifier(ch);
        } else {
          this.addToken(TokenType.ERROR, ch);
        }
        break;
    }
  }

  private isDigit(ch: string): boolean {
    return ch >= '0' && ch <= '9';
  }

  private isAlpha(ch: string): boolean {
    return (ch >= 'a' && ch <= 'z') ||
           (ch >= 'A' && ch <= 'Z') ||
           ch === '_';
  }

  private isAlphaNumeric(ch: string): boolean {
    return this.isAlpha(ch) || this.isDigit(ch);
  }

  private scanComment(): void {
    let value = '#';
    while (!this.isAtEnd() && this.peek() !== '\n') {
      value += this.advance();
    }
    this.addToken(TokenType.COMMENT, value);
  }

  private scanString(): void {
    let value = '';
    while (!this.isAtEnd() && this.peek() !== '"') {
      if (this.peek() === '\\' && this.peekNext() === '"') {
        this.advance();
        value += this.advance();
      } else {
        value += this.advance();
      }
    }
    if (this.isAtEnd()) {
      this.addToken(TokenType.ERROR, 'Unterminated string');
      return;
    }
    this.advance(); // Closing quote
    this.addToken(TokenType.STRING, value);
  }

  private scanNumber(first: string): void {
    let value = first;
    while (this.isDigit(this.peek())) {
      value += this.advance();
    }
    if (this.peek() === '.' && this.isDigit(this.peekNext())) {
      value += this.advance(); // .
      while (this.isDigit(this.peek())) {
        value += this.advance();
      }
    }
    // Scientific notation
    if (this.peek() === 'e' || this.peek() === 'E') {
      value += this.advance();
      if (this.peek() === '+' || this.peek() === '-') {
        value += this.advance();
      }
      while (this.isDigit(this.peek())) {
        value += this.advance();
      }
    }
    this.addToken(TokenType.NUMBER, value);
  }

  private scanIdentifier(first: string): void {
    let value = first;
    while (this.isAlphaNumeric(this.peek())) {
      value += this.advance();
    }
    const type = KEYWORDS[value.toLowerCase()] || TokenType.IDENTIFIER;
    this.addToken(type, value);
  }
}

/**
 * Tokenize DNA-Lang source code.
 */
export function tokenize(source: string): Token[] {
  return new Lexer(source).tokenize();
}

File: src/genesis/dna_lang/lexer.py
"""
DNA::}{::Lang Lexer

Tokenizes DNA-Lang source code into a stream of tokens for parsing.
The lexer recognizes the novel syntax elements of DNA-Lang including
gene definitions, evolution blocks, and consciousness expressions.
"""

from __future__ import annotations
from enum import Enum, auto
from dataclasses import dataclass
from typing import List, Optional, Iterator


class TokenType(Enum):
    """Token types for DNA-Lang."""
    
    # Literals
    INTEGER = auto()
    FLOAT = auto()
    STRING = auto()
    IDENTIFIER = auto()
    
    # Keywords
    ORGANISM = auto()
    GENE = auto()
    EVOLVE = auto()
    EMIT = auto()
    TRACK = auto()
    MEASURE = auto()
    MUTATE = auto()
    IF = auto()
    ELSE = auto()
    WHILE = auto()
    FOR = auto()
    IN = auto()
    RETURN = auto()
    TRUE = auto()
    FALSE = auto()
    NULL = auto()
    
    # Special DNA-Lang keywords
    CONSCIOUSNESS = auto()
    COHERENCE = auto()
    DECOHERENCE = auto()
    CCCE = auto()
    PSI = auto()
    LAMBDA_PHI = auto()
    GAMMA = auto()
    
    # Operators
    PLUS = auto()        # +
    MINUS = auto()       # -
    STAR = auto()        # *
    SLASH = auto()       # /
    CARET = auto()       # ^
    PERCENT = auto()     # %
    
    # Comparison
    EQ = auto()          # ==
    NEQ = auto()         # !=
    LT = auto()          # <
    GT = auto()          # >
    LEQ = auto()         # <=
    GEQ = auto()         # >=
    
    # Assignment and definition
    ASSIGN = auto()      # =
    DEFINE = auto()      # ::=
    ARROW = auto()       # ->
    FAT_ARROW = auto()   # =>
    
    # Delimiters
    LPAREN = auto()      # (
    RPAREN = auto()      # )
    LBRACE = auto()      # {
    RBRACE = auto()      # }
    LBRACKET = auto()    # [
    RBRACKET = auto()    # ]
    COMMA = auto()       # ,
    DOT = auto()         # .
    COLON = auto()       # :
    SEMICOLON = auto()   # ;
    
    # Special DNA symbols
    DNA_BIND = auto()    # }{
    INTEGRAL = auto()    # ∫
    PARTIAL = auto()     # ∂
    PSI_SYMBOL = auto()  # Ψ
    PHI_SYMBOL = auto()  # Φ
    GAMMA_SYMBOL = auto()  # Γ
    LAMBDA_SYMBOL = auto() # Λ
    XI_SYMBOL = auto()   # Ξ
    TAU_SYMBOL = auto()  # τ
    
    # Special
    NEWLINE = auto()
    EOF = auto()
    
    # Comments
    COMMENT = auto()


KEYWORDS = {
    "organism": TokenType.ORGANISM,
    "gene": TokenType.GENE,
    "evolve": TokenType.EVOLVE,
    "emit": TokenType.EMIT,
    "track": TokenType.TRACK,
    "measure": TokenType.MEASURE,
    "mutate": TokenType.MUTATE,
    "if": TokenType.IF,
    "else": TokenType.ELSE,
    "while": TokenType.WHILE,
    "for": TokenType.FOR,
    "in": TokenType.IN,
    "return": TokenType.RETURN,
    "true": TokenType.TRUE,
    "false": TokenType.FALSE,
    "null": TokenType.NULL,
    "consciousness": TokenType.CONSCIOUSNESS,
    "coherence": TokenType.COHERENCE,
    "decoherence": TokenType.DECOHERENCE,
    "ccce": TokenType.CCCE,
    "psi": TokenType.PSI,
    "lambda_phi": TokenType.LAMBDA_PHI,
    "gamma": TokenType.GAMMA,
}


SPECIAL_SYMBOLS = {
    "Ψ": TokenType.PSI_SYMBOL,
    "Φ": TokenType.PHI_SYMBOL,
    "Γ": TokenType.GAMMA_SYMBOL,
    "Λ": TokenType.LAMBDA_SYMBOL,
    "Ξ": TokenType.XI_SYMBOL,
    "τ": TokenType.TAU_SYMBOL,
    "∫": TokenType.INTEGRAL,
    "∂": TokenType.PARTIAL,
}


@dataclass
class Token:
    """
    A lexical token from DNA-Lang source.
    
    Attributes:
        type: The token type.
        value: The token value.
        line: Source line number.
        column: Source column number.
    """
    type: TokenType
    value: str
    line: int
    column: int
    
    def __repr__(self) -> str:
        return f"Token({self.type.name}, {self.value!r}, {self.line}:{self.column})"


class LexerError(Exception):
    """Exception raised for lexer errors."""
    
    def __init__(self, message: str, line: int, column: int) -> None:
        super().__init__(f"Lexer error at {line}:{column}: {message}")
        self.line = line
        self.column = column


class Lexer:
    """
    DNA::}{::Lang lexer.
    
    Tokenizes source code into a sequence of tokens. Supports the
    full DNA-Lang syntax including Unicode mathematical symbols.
    
    Example:
        >>> lexer = Lexer("organism Test { gene x ::= 1 }")
        >>> tokens = lexer.tokenize()
    """
    
    def __init__(self, source: str) -> None:
        """
        Initialize the lexer.
        
        Args:
            source: The source code to tokenize.
        """
        self.source = source
        self.pos = 0
        self.line = 1
        self.column = 1
        self.tokens: List[Token] = []
    
    @property
    def current_char(self) -> Optional[str]:
        """Get the current character or None if at end."""
        if self.pos >= len(self.source):
            return None
        return self.source[self.pos]
    
    def peek(self, offset: int = 1) -> Optional[str]:
        """Peek ahead in the source."""
        pos = self.pos + offset
        if pos >= len(self.source):
            return None
        return self.source[pos]
    
    def advance(self) -> Optional[str]:
        """Advance to the next character."""
        char = self.current_char
        if char is not None:
            self.pos += 1
            if char == '\n':
                self.line += 1
                self.column = 1
            else:
                self.column += 1
        return char
    
    def skip_whitespace(self) -> None:
        """Skip whitespace characters."""
        while self.current_char is not None and self.current_char in ' \t\r':
            self.advance()
    
    def skip_comment(self) -> None:
        """Skip a comment."""
        if self.current_char == '#':
            while self.current_char is not None and self.current_char != '\n':
                self.advance()
        elif self.current_char == '/' and self.peek() == '/':
            while self.current_char is not None and self.current_char != '\n':
                self.advance()
        elif self.current_char == '/' and self.peek() == '*':
            self.advance()  # /
            self.advance()  # *
            while self.current_char is not None:
                if self.current_char == '*' and self.peek() == '/':
                    self.advance()  # *
                    self.advance()  # /
                    break
                self.advance()
    
    def read_string(self) -> Token:
        """Read a string literal."""
        quote = self.current_char
        start_line = self.line
        start_col = self.column
        self.advance()  # Opening quote
        
        value = ""
        while self.current_char is not None and self.current_char != quote:
            if self.current_char == '\\':
                self.advance()
                escape_char = self.current_char
                if escape_char == 'n':
                    value += '\n'
                elif escape_char == 't':
                    value += '\t'
                elif escape_char == 'r':
                    value += '\r'
                elif escape_char == '\\':
                    value += '\\'
                elif escape_char == quote:
                    value += quote
                else:
                    value += escape_char or ''
                self.advance()
            else:
                value += self.current_char
                self.advance()
        
        if self.current_char is None:
            raise LexerError("Unterminated string", start_line, start_col)
        
        self.advance()  # Closing quote
        return Token(TokenType.STRING, value, start_line, start_col)
    
    def read_number(self) -> Token:
        """Read a numeric literal."""
        start_line = self.line
        start_col = self.column
        value = ""
        
        # Integer part
        while self.current_char is not None and self.current_char.isdigit():
            value += self.current_char
            self.advance()
        
        # Check for float
        if self.current_char == '.' and self.peek() is not None and self.peek().isdigit():
            value += self.current_char
            self.advance()
            while self.current_char is not None and self.current_char.isdigit():
                value += self.current_char
                self.advance()
            
            # Scientific notation
            if self.current_char in 'eE':
                value += self.current_char
                self.advance()
                if self.current_char in '+-':
                    value += self.current_char
                    self.advance()
                while self.current_char is not None and self.current_char.isdigit():
                    value += self.current_char
                    self.advance()
            
            return Token(TokenType.FLOAT, value, start_line, start_col)
        
        # Scientific notation for integers
        if self.current_char in 'eE':
            value += self.current_char
            self.advance()
            if self.current_char in '+-':
                value += self.current_char
                self.advance()
            while self.current_char is not None and self.current_char.isdigit():
                value += self.current_char
                self.advance()
            return Token(TokenType.FLOAT, value, start_line, start_col)
        
        return Token(TokenType.INTEGER, value, start_line, start_col)
    
    def read_identifier(self) -> Token:
        """Read an identifier or keyword."""
        start_line = self.line
        start_col = self.column
        value = ""
        
        while (self.current_char is not None and 
               (self.current_char.isalnum() or self.current_char == '_')):
            value += self.current_char
            self.advance()
        
        # Check for keywords
        token_type = KEYWORDS.get(value.lower(), TokenType.IDENTIFIER)
        return Token(token_type, value, start_line, start_col)
    
    def tokenize(self) -> List[Token]:
        """
        Tokenize the entire source.
        
        Returns:
            List of tokens.
        """
        self.tokens = []
        
        while self.current_char is not None:
            start_line = self.line
            start_col = self.column
            char = self.current_char
            
            # Whitespace
            if char in ' \t\r':
                self.skip_whitespace()
                continue
            
            # Newline
            if char == '\n':
                self.tokens.append(Token(TokenType.NEWLINE, '\n', start_line, start_col))
                self.advance()
                continue
            
            # Comments
            if char == '#' or (char == '/' and self.peek() in '/*'):
                self.skip_comment()
                continue
            
            # Strings
            if char in '"\'':
                self.tokens.append(self.read_string())
                continue
            
            # Numbers
            if char.isdigit():
                self.tokens.append(self.read_number())
                continue
            
            # Identifiers and keywords
            if char.isalpha() or char == '_':
                self.tokens.append(self.read_identifier())
                continue
            
            # Special Unicode symbols
            if char in SPECIAL_SYMBOLS:
                self.tokens.append(Token(SPECIAL_SYMBOLS[char], char, start_line, start_col))
                self.advance()
                continue
            
            # Two-character operators
            two_char = char + (self.peek() or '')
            if two_char == '::' and self.peek(2) == '=':
                self.tokens.append(Token(TokenType.DEFINE, '::=', start_line, start_col))
                self.advance()
                self.advance()
                self.advance()
                continue
            
            if two_char == '}{':
                self.tokens.append(Token(TokenType.DNA_BIND, '}{', start_line, start_col))
                self.advance()
                self.advance()
                continue
            
            if two_char == '->':
                self.tokens.append(Token(TokenType.ARROW, '->', start_line, start_col))
                self.advance()
                self.advance()
                continue
            
            if two_char == '=>':
                self.tokens.append(Token(TokenType.FAT_ARROW, '=>', start_line, start_col))
                self.advance()
                self.advance()
                continue
            
            if two_char == '==':
                self.tokens.append(Token(TokenType.EQ, '==', start_line, start_col))
                self.advance()
                self.advance()
                continue
            
            if two_char == '!=':
                self.tokens.append(Token(TokenType.NEQ, '!=', start_line, start_col))
                self.advance()
                self.advance()
                continue
            
            if two_char == '<=':
                self.tokens.append(Token(TokenType.LEQ, '<=', start_line, start_col))
                self.advance()
                self.advance()
                continue
            
            if two_char == '>=':
                self.tokens.append(Token(TokenType.GEQ, '>=', start_line, start_col))
                self.advance()
                self.advance()
                continue
            
            # Single-character tokens
            single_char_tokens = {
                '+': TokenType.PLUS,
                '-': TokenType.MINUS,
                '*': TokenType.STAR,
                '/': TokenType.SLASH,
                '^': TokenType.CARET,
                '%': TokenType.PERCENT,
                '<': TokenType.LT,
                '>': TokenType.GT,
                '=': TokenType.ASSIGN,
                '(': TokenType.LPAREN,
                ')': TokenType.RPAREN,
                '{': TokenType.LBRACE,
                '}': TokenType.RBRACE,
                '[': TokenType.LBRACKET,
                ']': TokenType.RBRACKET,
                ',': TokenType.COMMA,
                '.': TokenType.DOT,
                ':': TokenType.COLON,
                ';': TokenType.SEMICOLON,
            }
            
            if char in single_char_tokens:
                self.tokens.append(Token(single_char_tokens[char], char, start_line, start_col))
                self.advance()
                continue
            
            # Unknown character
            raise LexerError(f"Unexpected character: {char!r}", start_line, start_col)
        
        # Add EOF token
        self.tokens.append(Token(TokenType.EOF, '', self.line, self.column))
        return self.tokens
    
    def __iter__(self) -> Iterator[Token]:
        """Iterate over tokens."""
        if not self.tokens:
            self.tokenize()
        return iter(self.tokens)
